### ISSUE TITLE
Invalid XML for Jenkins?

### ISSUE THREAD

**Issue by kostajh** [n1k0/casperjs#824]
I'm getting an error with the results.xml file generated by CasperJS. I'm pointing `casper test` at a directory with four tests and the `results.xml` file is quite large.

When looking at a failed build, Jenkins is not telling me anything about successes and failures (other than it failed). The XML file isn't parsed properly.

Here is the output:

```
Failed to read test report file /var/lib/jenkins/workspace/Staging-Tests/tests/casperjs/results.xml
org.dom4j.DocumentException: Error on line 1 of document file:///var/lib/jenkins/workspace/Staging-Tests/tests/casperjs/results.xml : The entity "nbsp" was referenced, but not declared. Nested exception: The entity "nbsp" was referenced, but not declared.
	at org.dom4j.io.SAXReader.read(SAXReader.java:482)
	at org.dom4j.io.SAXReader.read(SAXReader.java:264)
	at hudson.tasks.junit.SuiteResult.parse(SuiteResult.java:123)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:273)
	at hudson.tasks.junit.TestResult.parsePossiblyEmpty(TestResult.java:229)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:164)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:147)
	at hudson.tasks.junit.TestResult.<init>(TestResult.java:123)
	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:117)
	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:90)
	at hudson.FilePath.act(FilePath.java:914)
	at hudson.FilePath.act(FilePath.java:887)
	at hudson.tasks.junit.JUnitParser.parse(JUnitParser.java:87)
	at hudson.tasks.junit.JUnitResultArchiver.parse(JUnitResultArchiver.java:121)
	at hudson.tasks.junit.JUnitResultArchiver.perform(JUnitResultArchiver.java:133)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:784)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:756)
	at hudson.model.Build$BuildExecution.post2(Build.java:183)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:705)
	at hudson.model.Run.execute(Run.java:1695)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)
	at hudson.model.ResourceController.execute(ResourceController.java:88)
	at hudson.model.Executor.run(Executor.java:231)
Caused by: org.xml.sax.SAXParseException; systemId: file:///var/lib/jenkins/workspace/Staging-Tests/tests/casperjs/results.xml; lineNumber: 1; columnNumber: 6680; The entity "nbsp" was referenced, but not declared.
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:198)
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:177)
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:441)
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:368)
	at com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1388)
	at com.sun.org.apache.xerces.internal.impl.XMLScanner.scanAttributeValue(XMLScanner.java:897)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanAttribute(XMLNSDocumentScannerImpl.java:436)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:253)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2717)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:607)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:116)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:489)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:835)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:764)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:123)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1210)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:568)
	at org.dom4j.io.SAXReader.read(SAXReader.java:465)
	... 23 more
Nested exception: 
org.xml.sax.SAXParseException; systemId: file:///var/lib/jenkins/workspace/Staging-Tests/tests/casperjs/results.xml; lineNumber: 1; columnNumber: 6680; The entity "nbsp" was referenced, but not declared.
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:198)
	at com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:177)
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:441)
	at com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:368)
	at com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1388)
	at com.sun.org.apache.xerces.internal.impl.XMLScanner.scanAttributeValue(XMLScanner.java:897)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanAttribute(XMLNSDocumentScannerImpl.java:436)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:253)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2717)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:607)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:116)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:489)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:835)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:764)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:123)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1210)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:568)
	at org.dom4j.io.SAXReader.read(SAXReader.java:465)
	at org.dom4j.io.SAXReader.read(SAXReader.java:264)
	at hudson.tasks.junit.SuiteResult.parse(SuiteResult.java:123)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:273)
	at hudson.tasks.junit.TestResult.parsePossiblyEmpty(TestResult.java:229)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:164)
	at hudson.tasks.junit.TestResult.parse(TestResult.java:147)
	at hudson.tasks.junit.TestResult.<init>(TestResult.java:123)
	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:117)
	at hudson.tasks.junit.JUnitParser$ParseResultCallable.invoke(JUnitParser.java:90)
	at hudson.FilePath.act(FilePath.java:914)
	at hudson.FilePath.act(FilePath.java:887)
	at hudson.tasks.junit.JUnitParser.parse(JUnitParser.java:87)
	at hudson.tasks.junit.JUnitResultArchiver.parse(JUnitResultArchiver.java:121)
	at hudson.tasks.junit.JUnitResultArchiver.perform(JUnitResultArchiver.java:133)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:784)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:756)
	at hudson.model.Build$BuildExecution.post2(Build.java:183)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:705)
	at hudson.model.Run.execute(Run.java:1695)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:46)
	at hudson.model.ResourceController.execute(ResourceController.java:88)
	at hudson.model.Executor.run(Executor.java:231)
```

**Comment by amerikan** [comment]
I'm having a similar issue with Jenkins.

**Comment by amerikan** [comment]
The XML seems fine too me. Not sure why is xUnit showing the ERROR? Here's the generated XML:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites xmlns="http://www.w3.org/1999/xhtml" time="13.016">
  <testsuite name="Settings Page" tests="12" failures="0" errors="0" time="13.016" timestamp="2016-02-26T23:53:08.535Z" package="tests/suite/settings">
    <testcase name="Log in with valid credentials successful" classname="tests/suite/settings" time="2.957"/>
    <testcase name="Invalidate empty first name" classname="tests/suite/settings" time="1.36"/>
    <testcase name="Invalidate empty last name" classname="tests/suite/settings" time="0"/>
    <testcase name="Can update first name" classname="tests/suite/settings" time="1.106"/>
    <testcase name="Can update last name" classname="tests/suite/settings" time="0"/>
    <testcase name="Success profile udpate message appears" classname="tests/suite/settings" time="0.08"/>
    <testcase name="Invalidate empty password" classname="tests/suite/settings" time="4.147"/>
    <testcase name="Invalidate password update when current password is wrong" classname="tests/suite/settings" time="0.982"/>
    <testcase name="Invalidate password update when new passwords don't match each other" classname="tests/suite/settings" time="1.119"/>
    <testcase name="Invalidate password update when new password is blank" classname="tests/suite/settings" time="0.04"/>
    <testcase name="Invalidate password update when re-type new passwords is blank" classname="tests/suite/settings" time="0.001"/>
    <testcase name="Password can be updated" classname="tests/suite/settings" time="1.224"/>
    <system-out/>
  </testsuite>
</testsuites>
```

**Comment by BIGjuevos** [comment]
I'm adopting resolution of this bug.  Please feel free to assign to me.

**Comment by BIGjuevos** [comment]
I have a [rough draft](https://github.com/BIGjuevos/casperjs/commit/d9f31e82e1bca5225c12f7ffc8f2266bb1cbee31) of this that outputs valid XML with all former text contents set as CDATA.  However, I need to fix the tests and add some new ones.

@mickaelandrieu @istr - Any early feedback would be appreciated.

**Comment by istr** [comment]
The implementation looks ok.

**Comment by mickaelandrieu** [comment]
@BIGjuevos can you make a pull request please ? this sounds ok to me too :)

**Comment by BIGjuevos** [comment]
@mickaelandrieu @istr I'm updating the tests, as the default document type is HTMLDocument which does not behave well with CDATA.  I'm having to update all of the tests to use an xmldocument as well.  I'll have a pull request up soon.

**Comment by BIGjuevos** [comment]
See #1499 @mickaelandrieu @istr @kostajh

**Comment by istr** [comment]
Closed by merged PR #1499. Thanks a lot for that PR @BIGjuevos
